### PR TITLE
Refactor rich text package to avoid using blocks packages as a dependcy

### PR DESCRIPTION
**Pull request by gziolo**
## Description
Supersedes: #11122, #11197.

> Currently @wordpress/blocks is not added as a dependency.

This PR removes this dependency completely by duplicating `normalizeIconObject` and `isValidIcon` logic in `@wordpress/rich-text`. It's based on the discussion in #11197: https://github.com/WordPress/gutenberg/pull/11197#discussion_r228878183.

> Or maybe just duplicate the function as normalizing block icons can have a different meaning than normalizing format icons.

It turns out, we still need to keep `normalizeIconObject` public as it is used to normalize icons passed to the inserter for Inline Elements.

## How has this been tested?
`npm test` & `npm run test-e2e`

Block icons have pretty good coverage in e2e test (props to @jorgefilipecosta).
You also install manually `Gutenberg Test Block Icons` plugin bundled with Gutenberg Docker instance and verify that all icons are properly rendered in all types of inserters.

## Types of changes
Refactoring.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

File: packages/rich-text/src/utils.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	return !! icon && (
+		isString( icon ) ||
+		isValidElement( icon ) ||
+		isFunction( icon ) ||
+		icon instanceof Component
+	);
+}
+
+/**
+ * Function that receives an icon as set by the blocks during the registration
```
Comment (gziolo) [inline packages/rich-text/src/utils.js]: `blocks` -> `format type`

File: packages/rich-text/src/utils.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ * and returns a new icon object that is normalized so we can rely on just on possible icon structure
+ * in the codebase.
+ *
+ * @param {(Object|string|WPElement)} icon  Slug of the Dashicon to be shown
+ *                                          as the icon for the block in the
```
Comment (gziolo) [inline packages/rich-text/src/utils.js]: Same as above.

File: packages/editor/src/components/rich-text/format-edit.js [hunk 3]
```diff
@@ -43,7 +44,7 @@ function FillInserterListItem( props ) {
 					return null;
 				}
 
-				return <InserterListItem { ...props } />;
+				return <InserterListItem { ...props } icon={ normalizeIconObject( props.icon ) } />;
```
Comment (youknowriad) [inline packages/editor/src/components/rich-text/format-edit.js]: Is the `normalizeIconObject` call still necessary here, shouldn't we call it on registration instead?

File: packages/rich-text/src/utils.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ *                                          inserter, or element or an object describing the icon.
+ *
+ * @return {Object} Object describing the icon.
+ */
+export function normalizeIconObject( icon ) {
+	if ( ! icon ) {
+		icon = 'block-default';
```
Comment (gziolo) [inline packages/rich-text/src/utils.js]: It probably should differ. @iseulde where do we use those icons for format types?

Comment (youknowriad) [inline packages/rich-text/src/utils.js]: Is this the default icon for richText formats? Shouldn't we use another icon instead?

File: packages/editor/src/components/rich-text/format-edit.js (hunk 3 above)
Comment (gziolo) [reply packages/editor/src/components/rich-text/format-edit.js]: We don't have access to `InserterListItem` unfortunately as it is part of the `edit` method.

**Comment by chrisvanpatten** [comment]
We have enough special icon specific code around Gutenberg that maybe it’s time to consider a `@wordpress/icons` package that consolidated some of these utilities? Not sure if it would help this case (only quickly scanned; I’m on mobile right now) but if it would remove duplication it might be nice.

**Comment by gziolo** [comment]
> We have enough special icon specific code around Gutenberg that maybe it’s time to consider a @wordpress/icons package that consolidated some of these utilities? Not sure if it would help this case (only quickly scanned; I’m on mobile right now) but if it would remove duplication it might be nice.

Yeah, definitely a good idea for a follow-up PR. It's just tons of work to create a new package. We want to stabilize Format API first and then do some more cleanup after 4.2 release is out.

File: packages/editor/src/components/rich-text/format-edit.js (hunk 3 above)
Comment (youknowriad) [reply packages/editor/src/components/rich-text/format-edit.js]: Not sure what's the relation with `InserterListItem`. I was just suggesting keeping this line `return <InserterListItem { ...props } />` here and calling `normalizeIconObject`inside `registerFormatType`

Comment (youknowriad) [reply packages/editor/src/components/rich-text/format-edit.js]: Maybe, I'm not understanding this properly, but this icon is supposed to be the format icon right?

Comment (gziolo) [reply packages/editor/src/components/rich-text/format-edit.js]: Nope, it comes from the component directly

Comment (youknowriad) [reply packages/editor/src/components/rich-text/format-edit.js]: Then, why do we even need to normalize it?

Comment (youknowriad) [reply packages/editor/src/components/rich-text/format-edit.js]: Regardless, shouldn't we import `normalizeIconObject` from `rich-text` here?

**Comment by gziolo** [comment]
@youknowriad I removed `icon` from the `settings` object used during a format type registration. It turns out it is never used in the code so it solves the issue for now as we avoid code duplication. As noted by you and @chrisvanpatten we should look into refactoring the existing code and extracting the problematic logic into its own package as soon as we introduce back `icon` property and start using it for format types. Related change: 203ae46

File: packages/rich-text/package.json [hunk 5]
```diff
@@ -22,9 +22,11 @@
 	"dependencies": {
 		"@babel/runtime": "^7.0.0",
 		"@wordpress/data": "file:../data",
+		"@wordpress/element": "file:../element",
```
Comment (youknowriad) [inline packages/rich-text/package.json]: We probably don't need the changes here anymore right?

**Review by youknowriad — APPROVED** [review]
LGTM 👍 Feel free to merge when the tests pass (and potentially remove the useless package.json changes)

**Pull request merged**